### DISCUSSION TITLE
Configuring sqlpage in Docker

### DISCUSSION THREAD

**Discussion by lozdown** [sqlpage/SQLPage#651] [Q&A]
I have a working sqlpage application that shows photos in a slideshow (the carusel component does not work well with large collections) on an obsolete ipad. This slideshow application uses the database of a digicam photo collection to select photos to display based on user input.
 
I want to run this application on a rpi using docker.

I have mapped the website *.sql etc files, and the sqlpage.json, however I cannot work out what database path to configure in the sqlpage.json. 

**Application filesystem (in $pwd) :**

```

├── sqlpage
│   ├── digikam4.db -> /home/laurie/Pictures/digikam4.db
│   ├── sqlpage.json
├── sqlpage.bin
└── sqlpage-root
│   ├── 1-Pictures -> /home/laurie/Documents/SAVTE/Library/1-Pictures/
│   ├── change-image.js
│   ├── foto-base -> /home/laurie/Pictures/foto-base/
│   ├── index.sql
│   ├── selection.sql
│   ├── slide-show.css
│   ├── picture-list.sql
│   ├── pre-index.sql
│   ├── json2.js
```

**sqlpage.json:**
```
{
   "database_url": "sqlite://sqlpage/digikam4.db?mode=ro",
     "web_root": "./sqlpage-root",
  "port": "8080"
}
```

**cli docker command**

```
docker run -it --name sqlpage -p 8080:8080 --volume "$(pwd)/sqlpage-root:/var/www" --volume "$(pwd)/sqlpage:/etc/sqlpage:ro" --rm lovasoa/sqlpage
[2024-10-22T14:29:39.951Z INFO  sqlpage::webserver::database::connect] Connecting to database: sqlite://sqlpage/digikam4.db?mode=ro
[2024-10-22T14:29:39.951Z WARN  sqlpage::webserver::database::connect] Failed to connect to the database: error returned from database: (code: 14) unable to open database file. Retrying in 5 seconds.
```
...

**Comment by lovasoa** [comment]
The database url to use is `sqlite:///etc/sqlpage/digikam4.db`

**Comment by lozdown** [comment]
Thank you for your very fast reply and suggestion. Unfortunately I had already tried it and it didn't work. 

I have now discovered that the problem lies with the symbolic link which redirects the database to a file elsewhere on my host file system. By copying the actual database to ($pwd)/sqlpage the database will then open. I assume that is because the link is working in docker filespace. 

This leaves me with the problem of how to make the picture files available to the application - I need to symbolic link them as I do not want to copy all the files (there are 30000+) . I guess I must be able to configure docker --volume option somehow (I am a complete novice with docker)

**Reply by lovasoa** [reply]
I'm not sure I understand your problem. Why don't you mount the original database file and the original picture folder into your docker container with `--volume` ? 
Mounting a file or a folder into a container does not involve any file copying; it's instantaneous whatever the size of what you are mounting.

**Reply by lozdown** [reply]
Sorry, I didn't provide the full problem specification. There are a number of extra constraints, in particular the digikam database contains the full path names for the pictures catalogue which are stored on an external drive.

The sqlpage application has to map the database paths to the target files. Now I understand better the use of --volume I can see how this can be done. Many thanks for your leads.

**Reply by freesocket** [reply]
I had the same problem when running with Podman.

So after several tries i tried with `-u root` and the problem went away.
